File: src/main/java/com/timepath/steam/io/storage/VPKPlugin.java
package com.timepath.steam.io.storage;

import com.timepath.vfs.SimpleVFile;
import com.timepath.vfs.provider.ProviderPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.kohsuke.MetaInfServices;

import java.io.File;
import java.io.IOException;
import java.util.Collection;

@MetaInfServices(ProviderPlugin.class)
public class VPKPlugin implements ProviderPlugin {

    @NotNull
    @Override
    public SimpleVFile.FileHandler register() {
        return new SimpleVFile.FileHandler() {
            @Override
            public Collection<SimpleVFile> handle(@NotNull File file) throws IOException {
                if (!file.getName().endsWith("_dir.vpk")) return null;
                VPK ar = VPK.Companion.loadArchive(file);
                if (ar == null) return null;
                return ar.list();
            }
        };
    }
}

File: src/main/kotlin/com/timepath/steam/io/storage/VPKPlugin.kt
package com.timepath.steam.io.storage

import com.timepath.vfs.SimpleVFile
import com.timepath.vfs.provider.ProviderPlugin
import org.kohsuke.MetaInfServices

import java.io.File

/**
 * @author TimePath
 */
MetaInfServices(javaClass<ProviderPlugin>())
public class VPKPlugin : ProviderPlugin {
    override fun register() = object : SimpleVFile.FileHandler {
        override fun handle(file: File): Collection<SimpleVFile>? {
            if (!file.getName().endsWith("_dir.vpk")) return null
            return VPK.loadArchive(file)!!.list()
        }
    }
}
